File: 3644/solve.cpp
#include <stdio.h>
#include <vector>

using namespace std;

int solutions = 0;
int target = -1;
int target_n = -1;
int current_lcm = -1;
int *scores = NULL;
int *valid_points = NULL;
vector<int> *step_cache = NULL;
vector<int> **conn = NULL;

int gcd(int a, int b)
{
  while(b)
  {
    int c = b;
    b = a % b;
    a = c;
  }

  return a;
}

int lcm(int a, int b)
{
  return (long long)a * b / gcd(a, b);
}

int scan_data()
{
  delete []scores;
  delete []valid_points;
  delete step_cache;

  for (int i = 0; i < target_n; ++i) {
    if (conn) {
      delete conn[i];
    }
  }

  delete []conn;

  int m;
  if (scanf("%d %d %d",&target_n, &m, &target) == EOF) {
    return 0;
  }

  scores = new int[target_n];
  valid_points = new int[target_n];
  step_cache = new vector<int>();
  conn = new vector<int>*[target_n];

  current_lcm = 1;

  for (int i = 0; i < target_n; ++i) {
    conn[i] = new vector<int>();
    step_cache->push_back(0);
  }

  for (int i = 0; i < m; ++i) {
    int f, t;
    scanf("%d %d", &f, &t);

    conn[--f]->push_back(--t);
  }

  for (int i = 0; i < target_n; ++i) {
    scanf("%d", scores + i);

    valid_points[i] = scores[i] == gcd(scores[i], target);
  }

  return 1;
}

int visit(int p)
{
  //printf("visit: %d\n", p);
  int value = lcm(current_lcm, scores[p]);
  step_cache->push_back(value);
  current_lcm = value;

  int l = step_cache->size();
  if ((*step_cache)[l-1] == (*step_cache)[l-2]) {
    return false;
  }

  //printf("value: %d\n", value);
  if (p == target_n-1) {
    if (value == target) {
      ++solutions;
    }

    return false;
  }

  return gcd(value, target) == value;
}

void rollback(int p)
{
  //printf("rollback: %d\n", p);
  step_cache->pop_back();
  current_lcm = (*step_cache)[step_cache->size()-1];
}

void graph_walk()
{
  int visited[target_n];
  int search_index[target_n];
  for (int i = 0; i < target_n; ++i) {
    visited[i] = false;
    search_index[i] = -1;
  }

  vector<int> path = vector<int>();

  int p = 0;
  while (true) {
    visited[p] = true;
    path.push_back(p);
    int valid = visit(p);
    int next_p = -1;

    if (valid) {
      vector<int> *candidates = conn[p];
      int ind = 0;
      while (candidates->size() > ind) {
        int c = candidates->at(ind);
        if (!visited[c] && valid_points[c]) {
          next_p = c;
          search_index[p] = ind;
          break;
        }
        else {
          ++ind;
        }
      }
    }

    if (valid && next_p >= 0) {
      p = next_p;
    }
    else {
      int rollback_done = false;
      visited[p] = false;
      path.pop_back();
      rollback(p);

      while (path.size() > 0) {
        int last_p = path[path.size()-1];
        vector<int> *candidates = conn[last_p];
        int ind = search_index[last_p] + 1;

        while (candidates->size() > ind) {
          int c = candidates->at(ind);
          if (!visited[c] && valid_points[c]) {
            search_index[last_p] = ind;
            p = c;
            rollback_done = true;
            break;
          }
        }

        if (rollback_done)
          break;
        else {
          search_index[last_p] = -1;
          visited[last_p] = false;
          path.pop_back();
          rollback(last_p);
        }
      }

      if (!rollback_done) {
        break;
      }
    }
  }
}

int main()
{
  while(scan_data()) {
    solutions = 0;
    graph_walk();

    printf("%d\n", solutions % 1000000007);
  }

  return 0;
}

File: 3644/solve.py
#!/usr/bin/env python

import os
import sys

solutions = 0
step_cache = None
current_lcm = None
target = None
scores = None
target_n = None
valid_points = None

def gcd(a, b):
  """Return greatest common divisor using Euclid's Algorithm."""
  while b:      
      a, b = b, a % b
  return a

def lcm(a, b):
  """Return lowest common multiple."""
  return a * b // gcd(a, b)

def visit(p):
  global solutions, current_lcm

  value = lcm(current_lcm, scores[p])
  step_cache.append(value)
  current_lcm = value

  #print "visit %d" % p
  #print scores[p]
  #print value
  #print target_n

  if step_cache[-1] == step_cache[-2]:
    return False

  if p == target_n:
    if value == target:
      solutions += 1

    return False

  if not valid_points[p]:
    return False

  return value < target

def rollback(p):
  global current_lcm

  step_cache.pop()
  current_lcm = step_cache[-1]

  #print "rollback %d" % p
  #print current_lcm

def graph_walk(conn):
  n = len(conn)
  visited = [False] * n
  path = []
  search_index = [-1] * n

  p = 0

  while True:
    visited[p] = True
    path.append(p)
    valid = visit(p)
    next_p = -1

    if valid:
      candidates = conn[p]
      ind = 0
      while len(candidates) > ind:
        c = candidates[ind][0]
        if not visited[c]:
          next_p = c
          search_index[p] = ind
          break
        else:
          ind += 1

    if valid and next_p >= 0:
      p = next_p
    else:
      rollback_done = False
      visited[p] = False
      path.pop()
      rollback(p)

      while len(path) > 0:
        last_p = path[-1]
        candidates = conn[last_p]
        ind = search_index[last_p] + 1

        while len(candidates) > ind:
          c = candidates[ind][0]
          if not visited[c]:
            search_index[last_p] = ind
            p = c
            rollback_done = True
            break

        if rollback_done:
          break
        else:
          search_index[last_p] = -1
          visited[last_p] = False
          path.pop()
          rollback(last_p)

      if not rollback_done:
        #print path
        #print visited
        #print search_index
        break

def build_data(lines, start):
  p = start
  n, m, k = lines[p].split()
  n = int(n)
  m = int(m)
  target = int(k)

  n_lines = m + 2
  conn = []
  for i in xrange(0,n):
    conn.append([])

  p += 1
  for i in xrange(0, m):
    f, t = lines[p].split()
    f = int(f) - 1
    t = int(t) - 1

    conn[f].append((t, 1))
    p += 1

  scores = []
  for s in lines[p].split():
    scores.append(int(s))

  return conn, scores, target, n_lines

def solve():
  lines = sys.stdin.readlines()
  p_lines = 0

  global scores, target, solutions, step_cache, target_n, current_lcm, valid_points

  while p_lines < len(lines):
    conn, scores, target, n_lines = build_data(lines, p_lines)
    solutions = 0
    step_cache = [0]
    target_n = len(conn) - 1
    current_lcm = 1

    valid_points = []
    for s in scores:
      valid_points.append(gcd(s, target) == s)

    graph_walk(conn)
    print solutions

    p_lines += n_lines

if __name__ == "__main__":
  solve()
